Index: core/src/main/java/hudson/model/Hudson.java
===================================================================
--- core/src/main/java/hudson/model/Hudson.java	(revision 23655)
+++ core/src/main/java/hudson/model/Hudson.java	(working copy)
@@ -2606,8 +2606,10 @@
      * Called once the user logs in. Just forward to the top page.
      */
     public void doLoginEntry( StaplerRequest req, StaplerResponse rsp ) throws IOException {
-        if(req.getUserPrincipal()==null)
+        if(req.getUserPrincipal() == null) {
             rsp.sendRedirect2("noPrincipal");
+            return;
+        }
 
         String from = req.getParameter("from");
         if(from!=null && from.startsWith("/") && !from.equals("/loginError")) {
