diff --git a/src/test/winstone/testCase/HttpConnectorTest.java b/src/test/winstone/testCase/HttpConnectorTest.java
index 0745d43..8798d28 100644
--- a/src/test/winstone/testCase/HttpConnectorTest.java
+++ b/src/test/winstone/testCase/HttpConnectorTest.java
@@ -8,6 +8,7 @@ package winstone.testCase;
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.Socket;
 import java.util.HashMap;
 import java.util.Map;
 
@@ -137,4 +138,41 @@ public class HttpConnectorTest extends TestCase {
         winstone.shutdown();
         Thread.sleep(500);
     }
+
+    public void testDOS() throws IOException, SAXException, InterruptedException {
+        Map args = new HashMap();
+        args.put("webroot", "target/testwebapp");
+        args.put("prefix", "/examples");
+        args.put("httpPort", "10006");
+        args.put("ajp13Port", "-1");
+        args.put("controlPort", "-1");
+        args.put("debug", "8");
+        args.put("logThrowingLineNo", "true");
+        Logger.init(Logger.FULL_DEBUG, System.out, true);
+        Launcher winstone = new Launcher(args);
+        for (int i = 0; i < 50; i++) {
+            final InputStream is = new Socket("localhost", 10006).getInputStream();
+            new Thread("DOS #" + i) {
+                @Override public void run() {
+                    try {
+                        is.read();
+                        while (true) {
+                            new Socket("localhost", 10006).getInputStream().read();
+                        }
+                    } catch (IOException x) {
+                        x.printStackTrace();
+                    }
+                }
+            }.start();
+        }
+        WebConversation wc = new WebConversation();
+        WebRequest wreq = new GetMethodWebRequest("http://localhost:10006/examples/CountRequestsServlet");
+        WebResponse wresp = wc.getResponse(wreq);
+        InputStream content = wresp.getInputStream();
+        assertTrue("Loading CountRequestsServlet", content.available() > 0);
+        content.close();
+        winstone.shutdown();
+        Thread.sleep(500);
+    }
+
 }
