Index: src/main/java/hudson/tasks/BuildWrapper.java
===================================================================
--- src/main/java/hudson/tasks/BuildWrapper.java	(revision 33626)
+++ src/main/java/hudson/tasks/BuildWrapper.java	(working copy)
@@ -31,6 +31,7 @@
 import hudson.model.Run.RunnerAbortedException;
 
 import java.io.IOException;
+import java.io.OutputStream;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Map;
@@ -174,6 +175,32 @@
     }
 
     /**
+     * Provides an opportunity for a {@link BuildWrapper} to decorate the {@link BuildListener} logger to be used by the build.
+     * 
+     * <p>
+     * This hook is called very early on in the build (even before {@link #setUp(AbstractBuild, Launcher, BuildListener)} is invoked.)
+     * 
+     * <p>
+     * The default implementation is no-op, which just returns the {@code logger} parameter as-is.
+     *
+     * @param build
+     *      The build in progress for which this {@link BuildWrapper} is called. Never null.
+     * @param logger
+     *      The default logger. Never null. This method is expected to wrap this logger.
+     *      This makes sure that when multiple {@link BuildWrapper}s attempt to decorate the same logger
+     *      it will sort of work.
+     * @return
+     *      Must not be null. If a fatal error happens, throw an exception.
+     * @throws RunnerAbortedException
+     *      If a fatal error is detected but the implementation handled it gracefully, throw this exception
+     *      to suppress stack trace.
+     * @since 1.372
+     */
+    public OutputStream decorateLogger(AbstractBuild build, OutputStream logger) throws IOException, InterruptedException, RunnerAbortedException {
+        return logger;
+    }
+
+    /**
      * {@link Action} to be displayed in the job page.
      *
      * @param job
Index: src/main/java/hudson/model/Run.java
===================================================================
--- src/main/java/hudson/model/Run.java	(revision 33626)
+++ src/main/java/hudson/model/Run.java	(working copy)
@@ -99,6 +99,8 @@
 import org.kohsuke.stapler.export.ExportedBean;
 
 import com.thoughtworks.xstream.XStream;
+import java.io.FileOutputStream;
+import java.io.OutputStream;
 
 import static java.util.logging.Level.FINE;
 
@@ -1236,8 +1238,21 @@
                 try {
                     Charset charset = Computer.currentComputer().getDefaultCharset();
                     this.charset = charset.name();
-                    listener = new StreamBuildListener(getLogFile(),charset);
 
+                    // don't do buffering so that what's written to the listener
+                    // gets reflected to the file immediately, which can then be
+                    // served to the browser immediately
+                    OutputStream logger = new FileOutputStream(getLogFile());
+                    RunT build = job.getBuild();
+                    if (project instanceof BuildableItemWithBuildWrappers && build instanceof AbstractBuild) {
+                        BuildableItemWithBuildWrappers biwbw = (BuildableItemWithBuildWrappers) project;
+                        for (BuildWrapper bw : biwbw.getBuildWrappersList()) {
+                            logger = bw.decorateLogger((AbstractBuild) build, logger);
+                        }
+                    }
+
+                    listener = new StreamBuildListener(logger,charset);
+
                     listener.started(getCauses());
 
                     RunListener.fireStarted(this,listener);
