Index: src/main/resources/hudson/model/Job/buildHistory.jelly
===================================================================
RCS file: /cvs/hudson/hudson/main/core/src/main/resources/hudson/model/Job/buildHistory.jelly,v
retrieving revision 1.1
diff -u -r1.1 buildHistory.jelly
--- src/main/resources/hudson/model/Job/buildHistory.jelly	5 Nov 2006 21:14:28 -0000	1.1
+++ src/main/resources/hudson/model/Job/buildHistory.jelly	10 Nov 2006 11:40:14 -0000
@@ -41,6 +41,14 @@
           </table>
         </td></tr>
       </j:if>
+      <j:if test="${!empty build.description}">
+        <tr>
+          <td></td>
+          <td style="padding:0;white-space:normal">
+            ${build.description}
+          </td>
+        </tr>
+      </j:if>
     </j:forEach>
     <tr class="build-row">
       <td colspan="2" align="right">
