diff --git a/src/java/winstone/ajp13/Ajp13IncomingPacket.java b/src/java/winstone/ajp13/Ajp13IncomingPacket.java
index 5a4ace4..ea6f1e4 100644
--- a/src/java/winstone/ajp13/Ajp13IncomingPacket.java
+++ b/src/java/winstone/ajp13/Ajp13IncomingPacket.java
@@ -160,7 +160,9 @@ public class Ajp13IncomingPacket {
 
             String attValue = readString (di,encoding);
 
-            this.attributes.put(attName, attValue);
+            if (attName != null && attValue != null) {
+                this.attributes.put(attName, attValue);
+            }
             Logger.log(Logger.FULL_DEBUG, Ajp13Listener.AJP_RESOURCES,
                     "Ajp13IncomingPacket.Attribute", attName,
                     attValue);
