Index: resources/WEB-INF/side-files/hudson/model/Run/console.jsp
===================================================================
RCS file: /cvs/hudson/hudson/resources/WEB-INF/side-files/hudson/model/Run/console.jsp,v
retrieving revision 1.7
diff -u -r1.7 console.jsp
--- resources/WEB-INF/side-files/hudson/model/Run/console.jsp	19 Jul 2006 04:44:13 -0000	1.7
+++ resources/WEB-INF/side-files/hudson/model/Run/console.jsp	24 Jul 2006 17:41:38 -0000
@@ -14,9 +14,12 @@
       <t:buildCaption>
         Console Output
       </t:buildCaption>
-    </td><td align="right">
+    </td>
+  </tr>
+  <tr>
+    <td>
       <a href="consoleText">
-        <img src="${rootURL}/images/24x24/document.gif" alt="">View as plain text
+        <img src="${rootURL}/images/24x24/document.gif" alt="Plain text">View as plain text
       </a>
     </td>
   </tr></table>
