Index: src/main/resources/hudson/model/Run/consoleFull.jelly
===================================================================
--- src/main/resources/hudson/model/Run/consoleFull.jelly	(revision 30088)
+++ src/main/resources/hudson/model/Run/consoleFull.jelly	(working copy)
@@ -21,7 +21,7 @@
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 -->
-<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+<st:compress xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
   <j:set var="consoleFull" value="true" />
   <st:include page="console.jelly" />
-</j:jelly>
+</st:compress>
