Index: main/core/src/main/java/hudson/model/AbstractProject.java
===================================================================
--- main/core/src/main/java/hudson/model/AbstractProject.java	(revision 10952)
+++ main/core/src/main/java/hudson/model/AbstractProject.java	(working copy)
@@ -1027,6 +1027,12 @@
             },
             req, rsp );
     }
+    
+    @Override
+    public void onCopiedFrom(Item src) {
+        super.onCopiedFrom(src);        
+        this.disabled = true; // disable a copied project by default
+    }
 
     /**
      * Finds a {@link AbstractProject} that has the name closest to the given name.
