diff --git a/core/src/main/resources/hudson/model/BuildTimelineWidget/control.jelly b/core/src/main/resources/hudson/model/BuildTimelineWidget/control.jelly
index c168f54..991accd 100644
--- a/core/src/main/resources/hudson/model/BuildTimelineWidget/control.jelly
+++ b/core/src/main/resources/hudson/model/BuildTimelineWidget/control.jelly
@@ -78,7 +78,6 @@ THE SOFTWARE.
                 intervalUnit:   Timeline.DateTime.DAY,
                 intervalPixels: 200,
                 eventSource:    eventSource1,
-                date:           d,
                 timeZone:       tz,
                 theme:          theme1,
                 layout:         'overview'  // original, overview, detailed
