--- hudson.orig 2009-09-14 09:13:56.000000000 -0700
+++ hudson      2009-09-14 09:13:35.000000000 -0700
@@ -65,7 +65,7 @@

     # --user in daemon doesn't prepare environment variables like HOME, USER, LOGNAME or USERNAME,
     # so we let su do so for us now
-    $SU $HUDSON_USER -c "$DAEMON $DAEMON_ARGS -- $JAVA $JAVA_ARGS -jar $HUDSON_WAR $HUDSON_ARGS" || return 2
+    $SU -s /bin/bash $HUDSON_USER -c "$DAEMON $DAEMON_ARGS -- $JAVA $JAVA_ARGS -jar $HUDSON_WAR $HUDSON_ARGS" || return 2
 }

 #