Index: ResponseImpl.java
===================================================================
--- ResponseImpl.java	(revision 1418)
+++ ResponseImpl.java	(working copy)
@@ -131,10 +131,11 @@
 
     public void serveExposedBean(StaplerRequest req, Object exposedBean, Flavor flavor) throws ServletException, IOException {
         String pad=null;
-        Writer w = getCompressedWriter(req);
 
         setContentType(flavor.contentType);
 
+        Writer w = getCompressedWriter(req);
+        
         if(flavor== Flavor.JSON) {
             pad = req.getParameter("jsonp");
             if(pad!=null) w.write(pad+'(');
