Index: src/main/java/hudson/model/Computer.java
===================================================================
--- src/main/java/hudson/model/Computer.java	(revision 26224)
+++ src/main/java/hudson/model/Computer.java	(working copy)
@@ -634,6 +636,9 @@
      */
     public final long getIdleStartMilliseconds() {
         long firstIdle = Long.MIN_VALUE;
+        for (Executor e : oneOffExecutors) {
+            firstIdle = Math.max(firstIdle, e.getIdleStartMilliseconds());
+        }
         for (Executor e : executors) {
             firstIdle = Math.max(firstIdle, e.getIdleStartMilliseconds());
         }
