Index: hudson/main/core/src/main/java/hudson/model/Job.java
===================================================================
--- hudson/main/core/src/main/java/hudson/model/Job.java	(revision 28661)
+++ hudson/main/core/src/main/java/hudson/model/Job.java	(working copy)
@@ -286,12 +286,13 @@
      * <p>
      * Much of Hudson assumes that the build number is unique and monotonic, so
      * this method can only accept a new value that's bigger than
-     * {@link #getNextBuildNumber()} returns. Otherwise it'll be no-op.
+     * {@link #getLastBuild()} returns. Otherwise it'll be no-op.
      * 
      * @since 1.199 (before that, this method was package private.)
      */
     public synchronized void updateNextBuildNumber(int next) throws IOException {
-        if (next > nextBuildNumber) {
+        RunT lastBuild = getLastBuild();
+        if (next > lastBuild.getNumber()) {
             this.nextBuildNumber = next;
             saveNextBuildNumber();
         }
