### Eclipse Workspace Patch 1.0
#P main
Index: war/resources/scripts/hudson-behavior.js
===================================================================
--- war/resources/scripts/hudson-behavior.js	(revision 28559)
+++ war/resources/scripts/hudson-behavior.js	(working copy)
@@ -641,14 +641,20 @@
     // structured form submission
     "FORM" : function(form) {
         crumb.appendToForm(form);
-        if(Element.hasClassName("no-json"))
+        if(Element.hasClassName(form, "no-json"))
             return;
         // add the hidden 'json' input field, which receives the form structure in JSON
         var div = document.createElement("div");
         div.innerHTML = "<input type=hidden name=json value=init>";
         form.appendChild(div);
         
-        form.onsubmit = function() { buildFormTree(this); };
+        var oldOnsubmit = form.onsubmit;
+        if(typeof oldOnsubmit=="function") {
+        	form.onsubmit = function() { buildFormTree(this); return oldOnsubmit.call(this); }
+        } else {
+        	form.onsubmit = function() { buildFormTree(this); };
+        }
+        
         form = null; // memory leak prevention
     },
 
