*** FormValidation.java	2010-09-11 09:50:42.000000000 +0200
--- FormValidation_patched.java	2010-09-13 17:16:18.000000000 +0200
***************
*** 228,233 ****
--- 228,238 ----
                          Stapler.getCurrentRequest().getContextPath()+ Hudson.RESOURCE_PATH+"/images/none.gif' height=16 width=1>"+
                          message+"</div>";
              }
+             
+             @Override
+             public String getMessage() {
+                 return message;
+             }
          };
      }
  
