Index: hudson/hudson/src/hudson/model/Hudson.java
===================================================================
RCS file: /cvs/hudson/hudson/src/hudson/model/Hudson.java,v
retrieving revision 1.85
diff -u -r1.85 Hudson.java
--- hudson/hudson/src/hudson/model/Hudson.java	14 Sep 2006 00:13:45 -0000	1.85
+++ hudson/hudson/src/hudson/model/Hudson.java	25 Sep 2006 17:49:47 -0000
@@ -149,8 +149,8 @@
         pluginManager = new PluginManager(context);
 
         load();
-        updateComputerList();
         if(slaves==null)    slaves = new ArrayList<Slave>();
+        updateComputerList();
     }
 
     /**
