Index: src/main/java/hudson/model/Computer.java
===================================================================
--- src/main/java/hudson/model/Computer.java	(revision 26224)
+++ src/main/java/hudson/model/Computer.java	(working copy)
@@ -591,7 +591,7 @@
     }
 
     public final int countExecutors() {
-        return executors.size();
+        return executors.size() + oneOffExecutors.size();
     }
 
     /**