Index: core/src/main/java/org/kohsuke/stapler/ClassDescriptor.java
===================================================================
--- core/src/main/java/org/kohsuke/stapler/ClassDescriptor.java	(revision 1057)
+++ core/src/main/java/org/kohsuke/stapler/ClassDescriptor.java	(working copy)
@@ -43,7 +43,7 @@
                     Class<?>[] p = m.getParameterTypes();
                     if(p.length==0)
                         continue;
-                    if(p[0].isAssignableFrom(clazz))
+                    if(!p[0].isAssignableFrom(clazz))
                         continue;
                     functions.add(new Function.StaticFunction(m).protectBy(m));
                 }
Index: core/src/main/java/org/kohsuke/stapler/MetaClass.java
===================================================================
--- core/src/main/java/org/kohsuke/stapler/MetaClass.java	(revision 1057)
+++ core/src/main/java/org/kohsuke/stapler/MetaClass.java	(working copy)
@@ -44,9 +44,15 @@
         this.webApp = webApp;
         this.baseClass = webApp.getMetaClass(clazz.getSuperclass());
         this.classLoader = MetaClassLoader.get(clazz.getClassLoader());
+        
+        Class<?>[] wrappers = new Class<?>[0];
+        WrapperManager wrapperManager = WrapperManager.get(webApp.context);
+        if (wrapperManager != null) {
+        	wrappers = (Class<?>[]) wrapperManager.getWrappers(clazz).toArray(new Class<?>[0]);
+        }
 
         buildDispatchers(
-            new ClassDescriptor(clazz,null/*support wrappers*/));
+            new ClassDescriptor(clazz, wrappers));
     }
 
     /**
