diff --git a/core/src/main/java/hudson/model/Computer.java b/core/src/main/java/hudson/model/Computer.java
index a846f60..eceafd6 100644
--- a/core/src/main/java/hudson/model/Computer.java
+++ b/core/src/main/java/hudson/model/Computer.java
@@ -905,9 +905,8 @@ public /*transient*/ abstract class Computer extends Actionable implements Acces
 
         // allow the administrator to manually specify the host name as a fallback. HUDSON-5373
         cachedHostName = channel.call(new GetFallbackName());
-
         hostNameCached = true;
-        return null;
+        return cachedHostName;
     }
 
     /**
